### ISSUE TITLE
GeneralOrderedMonoid should be renamed

### ISSUE THREAD

**Issue by DanGrayson** [Macaulay2/M2#1910]
`GeneralOrderedMonoid` should be changed to `OrderedFreeCommutativeMonoid`, and then we could add a new type, `OrderedFreeMonoid` to support the recent work by Mike and Frank on associative algebras.  Both types would be subtypes of `OrderedMonoid`.

@mikestillman

**Comment by DanGrayson** [comment]
Open question: should commutativity be a property, rather than being recorded in the type.

**Comment by mahrud** [comment]
What does `Ordered` mean here? Are there any examples of non-ordered monoids in M2? How about `Free`?

I propose following the algebraic hierarchy of an existing theorem prover, say Lean, to allow for a possible [interface](https://arxiv.org/abs/2101.07758) in the future. Based on [these notes](https://lean-forward.github.io/logical-verification/2018/42_notes.html), maybe we can start with something like this:

- `OrderedMonoid`: associative multiplication with 1
- `OrderedGroup`: associative, commutative addition with 0 and with additive inverse
- `OrderedRing`: `OrderedMonoid` and `OrderedGroup` with distributivity
- `OrderedCRing`: `OrderedRing` with commutativity
- `OrderedDRing`: `OrderedRing` with multiplicative inverse
- `OrderedField`: `OrderedCRing` with commutativity

These are just the ones that seem to be useful in M2. We can of course separate each to additive and multiplicative versions, with and without canceling or commutativity.

If everything is Ordered, we might as well drop the `Ordered` prefix.

A larger undertaking would be to slowly implement a "trait" system (or Concepts in C++ language) as opposed to a "type" system to make it easier to mix and match the traits above.

**Comment by mahrud** [comment]
@DanGrayson I plan on going ahead with this if there isn't anything else to add.

**Comment by DanGrayson** [comment]
Each polynomial ring comes with a degrees monoid, which harbors the degrees of the monomials.  I seem to recall hearing that sometimes it's useful to have degrees monoids with torson (@ggsmith @mikestillman), and degrees monoids don't have to be ordered.

I don't like the brevity of OrderedCRing and OrderedDRing, and as for OrderedRing, I think we don't have any ordered rings.  Similarly for OrderedField -- why bother with that?

As for dropping the "Ordered" prefix, that would be fine, as long as "OrderedMonoid" is retained as an ancestor.

Also, what's the point of OrderedGroup in our context?

**Comment by ggsmith** [comment]
Allowing the grading group of a polynomial ring to have torsion would be useful for some toric computations.

**Comment by mahrud** [comment]
> Allowing the grading group of a polynomial ring to have torsion would be useful for some toric computations.

My goal, roughly, is to make core functionalities more toric variety friendly, and working with Picard groups with torsion is definitely on my mind.

> I don't like the brevity of OrderedCRing and OrderedDRing

The short form comes from the category [CRing](https://ncatlab.org/nlab/show/CRing). I couldn't find a source that refers to `DRing` for division rings; we could call them `SkewField` instead.

> as for OrderedRing, I think we don't have any ordered rings

I intended the definition of `OrderedMonoid` to match what @mikestillman and @moorewf call 
a `FreeMonoid` and `OrderedRing` to match `FreeAlgebra`. Then `OrderedCRing` would be the class of commutative rings that are ubiquitous in M2.

The benefit of having `OrderedCRing` inherit from `OrderedRing` (rather than `NCRing` inherit from a commutative-by-default `Ring`) is that a lot of functions work identically regardless of commutativity.

> Similarly for OrderedField -- why bother with that?

Dealing with fields in M2 is kinda strange right now. Having a proper type for fields would simplify things like `isField` and `toField` (and their respective bugs).

> Also, what's the point of OrderedGroup in our context?

Dealing with groups is also very lacking in M2. The `degreesMonoid`, for instance, is literally a group. Packages like [InvariantRings](https://faculty.math.illinois.edu/Macaulay2/doc/Macaulay2/share/doc/Macaulay2/InvariantRing/html/_invariants_lp__Diagonal__Action_rp.html) and [GKMVarieties](https://faculty.math.illinois.edu/Macaulay2/doc/Macaulay2/share/doc/Macaulay2/GKMVarieties/html/___Example_co_spgeneralized_spflag_spvarieties.html) have implemented a lot of useful things with group actions, and yet there's no shared type for a group or group action to support commutative algebra involving those objects.

**Comment by DanGrayson** [comment]
> > I don't like the brevity of OrderedCRing and OrderedDRing
> 
> The short form comes from the category [CRing](https://ncatlab.org/nlab/show/CRing). I couldn't find a source that refers to `DRing` for division rings; we could call them `SkewField` instead.

CRing is not a good name -- it's too terse.  Why change what we have?

> 
> > as for OrderedRing, I think we don't have any ordered rings
> 
> I intended the definition of `OrderedMonoid` to match what @mikestillman and @moorewf call
> a `FreeMonoid` and `OrderedRing` to match `FreeAlgebra`. Then `OrderedCRing` would be the class of commutative rings that are ubiquitous in M2.

Those rings aren't ordered.

> 
> The benefit of having `OrderedCRing` inherit from `OrderedRing` (rather than `NCRing` inherit from a commutative-by-default `Ring`) is that a lot of functions work identically regardless of commutativity.

Yes, commutative rings should be a subclass of rings.

> 
> > Similarly for OrderedField -- why bother with that?
> 
> Dealing with fields in M2 is kinda strange right now. Having a proper type for fields would simplify things like `isField` and `toField` (and their respective bugs).

Those fields aren't ordered.

> 
> > Also, what's the point of OrderedGroup in our context?
> 
> Dealing with groups is also very lacking in M2. The `degreesMonoid`, for instance, is literally a group. Packages like [InvariantRings](https://faculty.math.illinois.edu/Macaulay2/doc/Macaulay2/share/doc/Macaulay2/InvariantRing/html/_invariants_lp__Diagonal__Action_rp.html) and [GKMVarieties](https://faculty.math.illinois.edu/Macaulay2/doc/Macaulay2/share/doc/Macaulay2/GKMVarieties/html/___Example_co_spgeneralized_spflag_spvarieties.html) have implemented a lot of useful things with group actions, and yet there's no shared type for a group or group action to support commutative algebra involving those objects.

Those groups aren't ordered, and they aren't ever used as the monoid of monomials for a monoid algebra.  It's probably premature to invent a type for groups, with no application in mind.

**Comment by mahrud** [comment]
What do you mean by "aren't ordered"? Can you give an example of something that is not-ordered in M2?
```m2
i1 : S = degreesMonoid{1,2,3};

i2 : T_0 < T_1

o2 = false

i3 : T_1 < T_0

o3 = true

i4 : T_1*T_2^-1 < T_0

o4 = false

i5 : T_0 < T_1*T_2^-1

o5 = true
```
Without an order, how can M2 even print anything?

**Comment by DanGrayson** [comment]
The (mathematical) orderings of interest are all on monoids, which serve as the monoids of monomials for our polynomial rings.  The polynomial rings are not ordered, especially since finite fields are never ordered rings.  For example, in this ring:
```m2

i11 : F = ZZ/3

o11 = F

i15 : 0_F ? 1_F

o15 = >

o15 : Keyword

i17 : 0_F + 2_F ? 1_F + 2_F

o17 = <

o17 : Keyword
```

**Comment by mahrud** [comment]
I think everywhere I mentioned `Ordered` I was referring to monomial orders, but you're right that I should drop `Ordered` from `OrderedGroup` because the additive group isn't ordered. You mentioned there are non-ordered monoids, is there an example of that?

**Comment by DanGrayson** [comment]
No, we don't have non-ordered monoids.  Eventually, though, if we have monoids with torsion, they won't be ordered.

**Comment by DanGrayson** [comment]
> Allowing the grading group of a polynomial ring to have torsion would be useful for some toric computations.

Such a grading monoid could probably be presented as a finitely generated free monoid modulo a (finitely generated) submonoid.  Is that right?  ( @ggsmith )

**Comment by DanGrayson** [comment]
Re: "Dealing with groups is also very lacking in M2. The degreesMonoid, for instance, is literally a group."

Yes, but being a group can be considered as a property of a monoid.  Why is it that for monomial ideals, which are currently a separate type of object, you propose removing their type and re-implementing them simply as ideals with the property of being monomial, whereas here, for monoids, you wish to create a separate type for monoids that have the property of being a group?  That seems inconsistent.

**Comment by ggsmith** [comment]
> Such a grading monoid could probably be presented as a finitely generated free monoid modulo a (finitely generated) submonoid. Is that right? ( @ggsmith )

Yes—I believe that gradings by finitely generated abelian groups would cover all toric applications.

**Comment by d-torrance** [comment]
> **Written by Claude** (Claude Opus 5, via Claude Code), not by @d-torrance, whose account posted it -- please weigh it accordingly.

Closing this: the rename has not happened and is not planned. `GeneralOrderedMonoid` is still
`m2/monoids.m2:188`, a direct subtype of `OrderedMonoid`, and no `OrderedFreeMonoid` was added —
the associative-algebras work this was meant to support went ahead without either change.

Two notes for anyone who reopens the question rather than the issue.

Somebody did start it. `AssociativeAlgebras.m2` has a `FreeMonoid` type and a `freeMonoid`
method under the comment `-- Toying with adding FreeMonoids`, but they sit below the file's
`end` line and so are never loaded; `isGlobalSymbol "FreeMonoid"` is false.

And the conflation this issue is about is load-bearing, not merely untidy:

```m2
i1 : class monoid (QQ[x,y])

o1 = GeneralOrderedMonoid

i2 : class monoid (QQ <| x,y |>)

o2 = GeneralOrderedMonoid
```

One type serves both the commutative and the free case, so renaming it to
`OrderedFreeCommutativeMonoid` today would misname the noncommutative one rather than finish
the job.

<sub>Triaged in a sweep of the open issues; source of truth: [`bug-triage/issues.tsv`](https://github.com/d-torrance/M2/blob/bug-triage/bug-triage/issues.tsv).</sub>
